### PR TITLE
[FLINK-32700] Support job drain for Savepoint upgrade mode

### PR DESCRIPTION
**Pull request by mananmangal**
## What is the purpose of the change

This pull request adds an option called `kubernetes.operator.job.drain-on-savepoint-deletion` to indicate whether a job should be drained before deleting a FlinkDeployment or FlinkSessionJob, only if savepoint on deletion is enabled.

## Brief change log

  - Add new configurable option to drain a job before deletion, if the savepoint on deletion is enabled

## Verifying this change

This change added tests and can be verified as follows:

  - Added unit tests with the new configuration option enabled `testSubmitAndDrainOnCleanUpWithSavepoint` in `ApplicationReconcilerTest` and `SessionJobReconcilerTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): No
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: yes
  - Core observer or reconciler logic that is regularly executed: yes

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconcilerTest.java [hunk 1]
```diff
@@ -1070,4 +1070,33 @@ public void testUpgradeReconciledGeneration() throws Exception {
...
+        reconciler.cleanup(deployment, TestUtils.createContextWithReadyFlinkDeployment());
+        assertEquals(
+                "savepoint_0",
+                deployment
+                        .getStatus()
+                        .getJobStatus()
+                        .getSavepointInfo()
+                        .getLastSavepoint()
+                        .getLocation());
+    }
```
Comment (gyfora) [inline flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconcilerTest.java:1101]: These tests you added don't actually cover that draining was used (the proper client method was called).

I suggest removing what you added and replacing with 2 tests in the `AbstractFlinkServiceTests` there you can test the changed service methods directly and make sure that the Flink client is called with the proper argument.

**Review by gyfora — CHANGES_REQUESTED** [review]

File: flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconcilerTest.java (hunk 1 above)
Comment (mananmangal) [reply flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconcilerTest.java:1101]: I understand, I will make these changes.

Comment (mananmangal) [reply flink-kubernetes-operator/src/test/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconcilerTest.java]: Replaced unit tests to AbstractFlinkServiceTests

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java [hunk 2]
```diff
@@ -185,6 +186,9 @@ public static FlinkOperatorConfiguration fromConfiguration(Configuration operato
         boolean savepointOnDeletion =
                 operatorConfig.get(KubernetesOperatorConfigOptions.SAVEPOINT_ON_DELETION);
 
+        boolean drainJobOnSavepointDeletion =
+                operatorConfig.get(KubernetesOperatorConfigOptions.DRAIN_ON_SAVEPOINT_DELETION);
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java:190]: Similar to this fix PR: https://github.com/apache/flink-kubernetes-operator/pull/659

These configs should not be part of the `FlinkOperatorConfiguration` which will prevent it for setting it for a per-resource level. It should be accessed from the observeConfig directly.

**Review by gyfora — CHANGES_REQUESTED** [review]

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java (hunk 2 above)
Comment (mananmangal) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/FlinkOperatorConfiguration.java]: Got it!
Updated the PR to reflect these changes.

**Comment by gyfora** [comment]
Looks good, please rebase on main and I will merge this tomorrow

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 3]
```diff
@@ -566,4 +566,12 @@ public static String operatorConfigKey(String key) {
...
                     .withDescription(
                             "Indicate whether a savepoint must be taken when deleting a FlinkDeployment or FlinkSessionJob.");
+
+    @Documentation.Section(SECTION_DYNAMIC)
+    public static final ConfigOption<Boolean> DRAIN_ON_SAVEPOINT_DELETION =
+            operatorConfig("job.drain-on-savepoint-deletion")
+                    .booleanType()
+                    .defaultValue(false)
+                    .withDescription(
+                            "Indicate whether a job should be drained before deleting a FlinkDeployment or FlinkSessionJob, only if savepoint on deletion is enabled.");
```
[the remark this answers is not in the document]
Comment (mananmangal) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java]: updated the description.

**Comment by mananmangal** [comment]
Looks like something went wrong during rebase, let me create another request and we can close this one.

**Comment by mananmangal** [comment]
#661  Created.
Closing this one due to incorrect rebase.

**Pull request closed without merging**